docs/architecture.rst

```diff
@@ -8,13 +8,93 @@ Endpoint nodes handle all notification PUT requests, looking up in DynamoDB to
 see what Push server the UAID is connected to. The Endpoint nodes then attempt
 delivery to the Push server.
 
-Push server's accept websocket connections (this can easily be HTTP/2 for
-WebPush), and deliver notifications to connected clients. They check DynamoDB
-for missed notifications as necessary.
+Push connection nodes accept websocket connections (this can easily be HTTP/2
+for WebPush), and deliver notifications to connected clients. They check
+DynamoDB for missed notifications as necessary.
 
 There will be many more Push servers to handle the connection node, while more
 Endpoint nodes can be handled as needed for notification throughput.
 
+Table Rotation
+==============
+
+To avoid costly table scans, autopush uses a rotating message and router table.
+Clients that haven't connected in 30-60 days will have heir router and message
+table entries will dropped and need to re-register.
+
+Tables are post-fixed with the year/month they are meant for, ie:
+
+    messages-2015-02
+
+Tables must be created and have their read/write units properly allocated by a
+separate process in advance of the month switch-over as autopush nodes will
+assume the tables already exist. Scripts are provided that can be run weekly to
+ensure all necessary tables are present, and tables old enough are dropped.
+
+Within a few days of the new month, the load on the prior months table will fall
+as clients transition to the new table. The read/write units on the prior
+month may then be lowered.
+
+Message Table
+-------------
+
+Due to the complexity of having notifications spread across two tables, several
+rules are used to avoid losing messages during the month transition.
+
+The logic for connection nodes is more complex, since only the connection node
+knows when the client connects, and how many messages it has read through.
+
+A new field will be added to the router table to indicate the last month the
+client has read connections through. This is independent of the last_connected
+since it is possible for a client to connect, fail to read its notifications,
+then reconnect. This field is updated for a new month when the client connects
+**after** it has ack'd all the notifications out of the last month.
+
+To avoid issues with time synchronization, the node the client is connected to
+acts as the source of truth for when the month has flipped over. Clients are
+only moved to the new table on connect, and only after reading/acking all the
+notifications for the prior month.
+
+**Rules for Endpoints**
+
+1. Check the router table to see the current_month the client is on.
+2. Read the chan list entry from the appropriate month message table to see if
+   its a valid channel.
+
+   If its valid, move to step 3.
```

Comment:
If the channel isn't valid, then the behavior is identical to current behavior, ie, the channel is unregistered and the message is 404'd.